Feature/nullable support utility

bigboybamo opened the pull request:
Made changes to files under `Utility` Library

bigboybamo committed:
First push for nullable

bigboybamo committed:
Fixed
-AsynHttpRequest
-LinuxPowerSupply
-MacOsPowerSupply

bigboybamo committed:
Fixed:
-BlockinQueue
-CallContext
-CommandLine
-DirectStreamLink

kenkendk commented on Duplicati/Library/Utility/AsyncHttpRequest.cs:
```diff
@@ -158,7 +158,7 @@ public Stream GetRequestStream(long contentlength = -1)
 
             if (m_state == RequestStates.GetRequest)
-                return (Stream)m_asyncRequest.GetResponseOrStream();
+                return (Stream)m_asyncRequest!.GetResponseOrStream();
```
Can we check if this is null and throw instead of suppress?

kenkendk commented on Duplicati/Library/Utility/AsyncHttpRequest.cs:
```diff
@@ -176,7 +176,7 @@ public Stream GetRequestStream(long contentlength = -1)
         {
             if (m_state == RequestStates.GetResponse)
-                return (WebResponse)m_asyncResponse.GetResponseOrStream();
+                return (WebResponse)m_asyncResponse!.GetResponseOrStream();
```
Can we check if this is null and throw instead of suppress?

kenkendk commented on Duplicati/Library/Utility/BlockingQueue.cs:
```diff
@@ -91,7 +91,8 @@ public object Current
                 {
-                    return m_current;
+
+                    return m_current!;
```
Are we sure this cannot be null?

kenkendk commented on Duplicati/Library/Utility/CallContextSettings.cs:
```diff
@@ -253,13 +253,13 @@ public static T Settings
                 var key = ContextID;
                 if (string.IsNullOrWhiteSpace(key))
-                    return default(T);
+                    return default(T)!;
```
Can this not be null?

kenkendk commented on Duplicati/Library/Utility/CallContextSettings.cs:
```diff
@@ -277,11 +277,11 @@ public static T Settings
             contextSettingsType = settingsKey;
             var res = new ContextGuard();
-            Settings = initial;
+            Settings = initial!;
```
Can this be done better? If we allow it to be null, then `Settings` should be nullable as well?

kenkendk commented on Duplicati/Library/Utility/CallContextSettings.cs:
```diff
@@ -253,13 +253,13 @@ public static T Settings
                     lock (_lock) // if not present but context id is not null, wait
                         if (!_settings.TryGetValue(key, out res))
-                            return default(T);
+                            return default(T)!;
```
This one looks like it could be null

kenkendk commented on Duplicati/Library/Utility/BlockingQueue.cs:
```diff
@@ -217,7 +218,7 @@ public T Dequeue()
 
                     if (m_completed)
-                        return default(T);
+                        return default(T)!;
```
This could technically return `null`?

kenkendk commented on Duplicati/Library/Utility/DirectStreamLink.cs:
```diff
@@ -252,7 +252,7 @@ private void readerClosed()
                 m_readerClosed = true;
                 m_signalBufferAvailable.Set(); // unblock potentially waiting writer
-                m_readerStream = null;
+                m_readerStream = null!;
```
It guess it should be nullable if we set it to null?

kenkendk commented on Duplicati/Library/Utility/DirectStreamLink.cs:
```diff
@@ -266,7 +266,7 @@ private void writerClosed()
                 m_writerClosed = true;
                 m_signalDataAvailable.Set(); // unblock potentially waiting reader
-                m_writerStream = null;
+                m_writerStream = null!;
```
It guess it should be nullable if we set it to null?

kenkendk commented on Duplicati/Library/Utility/Power/LinuxPowerSupplyState.cs:
```diff
@@ -57,12 +57,12 @@ private bool IsAC()
                     if (!reply)
                     {
-                        string sourceType = File.ReadLines(Path.Combine(source, "type")).FirstOrDefault();
+                        string? sourceType = File.ReadLines(Path.Combine(source, "type")).FirstOrDefault();
```
Feel free to use `var` :)

kenkendk commented on Duplicati/Library/Utility/Power/MacOSPowerSupplyState.cs:
```diff
@@ -52,7 +52,7 @@ private PowerSupply.Source GetSourcePmset()
 
                 var pi = System.Diagnostics.Process.Start(psi);
-                pi.WaitForExit(1000);
+                pi!.WaitForExit(1000);
```
maybe add:
```
var pi = System.Diagnostics.Process.Start(psi) ?? throw new Exception("Failed to start process");
```

kenkendk commented on Duplicati/Library/Utility/Power/MacOSPowerSupplyState.cs:
```diff
@@ -89,7 +89,7 @@ private PowerSupply.Source GetSourceIoreg()
 
                 var pi = System.Diagnostics.Process.Start(psi);
-                pi.WaitForExit(1000);
+                pi!.WaitForExit(1000);
```
maybe add:
```
var pi = System.Diagnostics.Process.Start(psi) ?? throw new Exception("Failed to start process");
```

bigboybamo committed:
Fixed continued
stopped ReadLimitLengthStream

bigboybamo committed:
Configured:
-SslCertificateValidator
-TempFile
-TempFolder
-ThrottledStream
-Uri
-UrlUtility
-WorkerThread

bigboybamo commented:
Hi @kenkendk I updated the PR again

Jojo-1000 commented on Duplicati/Library/Utility/CallContextSettings.cs:
```diff
@@ -277,11 +276,11 @@ public static T Settings
             contextSettingsType = settingsKey;
             var res = new ContextGuard();
-            Settings = initial;
+            Settings = initial ?? throw new ArgumentNullException(nameof(initial), "The initial value cannot be null.");
```
This is a contradiction. `default(T)` is null for any reference type.

bigboybamo replied:
Made a fox for this as well

Jojo-1000 commented on Duplicati/Library/Utility/CallContextSettings.cs:
```diff
@@ -236,7 +236,7 @@ public static class CallContextSettings<T>
         /// to prevent overwriting each others settings.
         /// </summary>
-        private static string contextSettingsType = null;
+        private static string? contextSettingsType = null;
```
This should not be nullable, because it can always be initialized in the constructor.

bigboybamo replied:
Fixed this

kenkendk commented on Duplicati/Library/Utility/FileBackedList.cs:
```diff
@@ -36,13 +36,13 @@ public abstract class FileBackedList<T> : IEnumerable<T>, IDisposable
         private class StreamEnumerator : IEnumerator<T>, System.Collections.IEnumerator
         {
-            private Stream m_stream;
+            private Stream? m_stream;
```
The stream can actually not be null and should be readonly.
The reason why it appears nullable is that it is being used as a `m_isDisposed` flag when null, which is bad practice.

bigboybamo replied:
marking this stream as readonly means it cannot be reassigned in this Dispose methosd
<img width="362" alt="image" src="https://github.com/duplicati/duplicati/assets/60369677/1920a11a-e050-4d83-982c-f7f2267782b1">

kenkendk commented on Duplicati/Library/Utility/FileBackedList.cs:
```diff
@@ -101,8 +107,8 @@ public void Reset()
-        private Library.Utility.TempFile m_file;
-        private Stream m_stream;
+        private Library.Utility.TempFile? m_file;
+        private Stream? m_stream;
```
Sam as above, it cannot be null and should be readonly.

bigboybamo replied:
Same thing here
<img width="370" alt="image" src="https://github.com/duplicati/duplicati/assets/60369677/d4a8cd0b-3830-42c3-a52b-d19bb6932c33">

So I'm not entirely sure how to proceed

kenkendk replied:
I see, it is lazy initialized, so we cannot make it readonly :'(

kenkendk commented on Duplicati/Library/Utility/JoinedFilterExpression.cs:
```diff
@@ -73,9 +75,9 @@ public static IFilter Join(IFilter first, IFilter second)
         {
             if (this.First.Empty)
-                return this.Second.ToString();
+                return this.Second.ToString() ?? throw new InvalidOperationException("Second is null");
```
This looks off. Can `.ToString()` ever be null? The `First` and `Second` properties are never null.

kenkendk commented on Duplicati/Library/Utility/OverrideableStream.cs:
```diff
@@ -33,7 +33,7 @@ namespace Duplicati.Library.Utility
     public class OverrideableStream : Stream
     {
-        protected System.IO.Stream m_basestream;
+        protected System.IO.Stream? m_basestream;
```
This one can never be null and should be `readonly`, it is being used as an `m_isDisposed` flag.
This removes all the other null checks.

bigboybamo replied:
Changing that stream to readonly means the assignment here causes an error
<img width="230" alt="image" src="https://github.com/duplicati/duplicati/assets/60369677/8f72559d-9b13-452a-a2af-fba61e450078">

kenkendk commented on Duplicati/Library/Utility/SslCertificateValidator.cs:
```diff
@@ -31,10 +31,16 @@ public class SslCertificateValidator
+            {
+                get
+                {
+                    return m_certificate ?? throw new InvalidOperationException("Certificate is null");
```
There is a check for this property in `RemoteOperation.cs`, I think this property should allow null to be returned.

kenkendk commented on Duplicati/Library/Utility/TempFile.cs:
```diff
@@ -35,7 +35,7 @@ public class TempFile : IDisposable
         public static string APPLICATION_PREFIX = Utility.getEntryAssembly().FullName.Substring(0, 3).ToLower(System.Globalization.CultureInfo.InvariantCulture) + "-";
 
-        private string m_path;
+        private string? m_path;
```
This is being set to null to avoid it being read after being disposed. It is better to make it `readonly` and add a `m_isDisposed` flag.

kenkendk commented on Duplicati/Library/Utility/TempFolder.cs:
```diff
@@ -29,7 +29,7 @@ namespace Duplicati.Library.Utility
     public class TempFolder : IDisposable
     {
-        private string m_folder;
+        private string? m_folder;
```
This should be `readonly` and a `m_isDisposed` flag should be added.

bigboybamo replied:
Not sure where to add the m_isDisposed flag?

kenkendk commented on Duplicati/Library/Utility/ThrottledStream.cs:
```diff
@@ -135,6 +135,7 @@ public override int Read(byte[] buffer, int offset, int count)
 				var chunksize = (int)Math.Min(remaining, m_readspeed <= 0 ? remaining : m_readspeed * 2);
 				DelayIfRequired(ref m_readspeed, chunksize, ref m_last_read_sample, ref m_current_read_counter, ref m_current_read_speed);
 
+				if(m_basestream == null) throw new InvalidOperationException("m_basestream is null");
```
This will go away once the stream is not nullable in `OverrideableStream`

kenkendk commented on Duplicati/Library/Utility/ThrottledStream.cs:
```diff
@@ -164,6 +165,7 @@ public override void Write(byte[] buffer, int offset, int count)
                 UpdateLimits();
 				var chunksize = (int)Math.Min(count, m_writespeed <= 0 ? count : m_writespeed * 2);
 				DelayIfRequired(ref m_writespeed, chunksize, ref m_last_write_sample, ref m_current_write_counter, ref m_current_write_speed);
+				if (m_basestream == null) throw new InvalidOperationException("m_basestream is null");
```
This will go away once the stream is not nullable in `OverrideableStream`

kenkendk commented on Duplicati/Library/Utility/WorkerThread.cs:
```diff
@@ -254,7 +254,7 @@ public Tx CurrentTask
             get
             {
-                return m_currentTask;
+                return m_currentTask ?? throw new InvalidOperationException("m_currentTask is null");
```
The return value can be null. As far as I can read the code, it starts out being null.

kenkendk reviewed (changes requested):
Sorry for the delay in reviewing, it is a much harder change in this project than what I anticipated.

I think some of the changes may need to be pushed to another PR if you feel they are too involved.

bigboybamo committed:
Make Corrections based on PR review

kenkendk commented on Duplicati/Library/Utility/SslCertificateValidator.cs:
```diff
@@ -31,10 +31,16 @@ public class SslCertificateValidator
             private readonly SslPolicyErrors m_errors = SslPolicyErrors.None;
 
-            public string Certificate { get { return m_certificate; } }
+            public string Certificate 
```
Should this not be `string?` as it returns `m_certificate` ?

kenkendk commented on Duplicati/Library/Utility/OverrideableStream.cs:
Can we make `protected readonly Stream m_basestream;`  ?
This requires only that the dispose call does not set it to null, and removes all the extra code for handling nullable.
